core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -281,7 +297,7 @@ public UpdateSchema updateColumn(String name, Type.PrimitiveType newType) {
     }
 
     Preconditions.checkArgument(
-        TypeUtil.isPromotionAllowed(field.type(), newType),
+        TypeUtil.isPromotionAllowed(field.type(), newType, formatVersion),
```

Comment:
You need to add Partition Field Validation too. [According to spec](https://iceberg.apache.org/spec/#:~:text=Type%20promotion%20is,timestamp%20or%20timestamp_ns)  `Type promotion is not allowed for a field that is referenced by source-id or source-ids of a partition field if the partition transform would produce a different value after promoting the type.`